consensus/polybft/polybft.go

```diff
@@ -0,0 +1,662 @@
+			return false
+		}
+
+		if err := p.syncer.Sync(nullHandler); err != nil {
+			panic(fmt.Errorf("failed to sync blocks. Error: %w", err))
+		}
+	}()
+
+	return nil
+}
+
+// startSealing is executed if the PolyBFT protocol is running in sealing mode.
+func (p *Polybft) startSealing() error {
+	p.logger.Info("Using signer", "address", p.key.String())
+
+	if err := p.startRuntime(); err != nil {
+		return fmt.Errorf("consensus runtime start failed: %w", err)
+	}
+
+	go func() {
+		// start the pbft process
+		p.startPbftProcess()
+	}()
+
+	return nil
+}
+
+// initializeConsensusConfig populates consensus configuration
+func (p *Polybft) initializeConsensusConfig() {
+	customConfigGeneric := p.config.Config.Config
+
+	var polyBftConfig PolyBFTConfig
+
+	customConfigJSON, _ := json.Marshal(customConfigGeneric)
+	json.Unmarshal(customConfigJSON, &polyBftConfig)
+
+	p.logger.Info(fmt.Sprintf("Polybft config %+v", polyBftConfig))
+
+	// TODO: Bridge, validators configuration
+	p.consensusConfig = &polyBftConfig
+}
+
+// startRuntime starts consensus runtime
+func (p *Polybft) startRuntime() error {
+	runtimeConfig := &runtimeConfig{
+		PolyBFTConfig: p.consensusConfig,
+		Key:           p.key,
+		DataDir:       p.dataDir,
+		Transport: &bridgeTransportWrapper{
+			topic:  p.bridgeTopic,
+			logger: p.logger.Named("bridge_transport"),
+		},
+		State:          p.state,
+		blockchain:     p.blockchain,
+		polybftBackend: p,
+	}
+
+	runtime, err := newConsensusRuntime(p.logger, runtimeConfig)
+	if err != nil {
+		return err
+	}
+
+	p.runtime = runtime
+
+	if runtime.IsBridgeEnabled() {
+		err := p.bridgeTopic.Subscribe(func(obj interface{}, from peer.ID) {
+			msg, _ := obj.(*proto.TransportMessage)
+			var transportMsg *TransportMessage
+			if err := json.Unmarshal(msg.Data, &transportMsg); err != nil {
+				panic(err)
+			}
+			if _, err := p.runtime.deliverMessage(transportMsg); err != nil {
+				p.logger.Warn(fmt.Sprintf("Failed to deliver message. Error: %s", err))
+			}
+		})
+		if err != nil {
+			return fmt.Errorf("topic subscription failed:%w", err)
+		}
+	}
+
+	return nil
+}
+
+func (p *Polybft) startPbftProcess() {
+	// wait to have at least n peers connected. The 2 is just an initial heuristic value
+	// Most likely we will parametrize this in the future.
+	if !p.waitForNPeers() {
+		return
+	}
+
+	// subscribe to new block events
+	var (
+		newBlockSub   = p.blockchain.SubscribeEvents()
+		syncerBlockCh = make(chan uint64)
+	)
+
+	// Receive a notification every time syncer manages
+	// to insert a valid block.
+	go func() {
+		eventCh := newBlockSub.GetEventCh()
+
+		for {
+			select {
+			case ev := <-eventCh:
+				currentBlockNum := p.blockchain.CurrentHeader().Number
+				if ev.Source == "syncer" {
+					if ev.NewChain[0].Number < currentBlockNum {
+						continue
+					}
+				}
+
+				if p.isSynced() {
+					syncerBlockCh <- currentBlockNum
+				}
+
+			case <-p.closeCh:
+				return
+			}
+		}
+	}()
+
+	defer newBlockSub.Close()
+
+SYNC:
```

Comment:
We shouldn't be using `goto` directives in Go in 2022, because we couldn't figure out how to write the program flow better.

Please refactor this method flow to not rely on goto label directives.

cc @vcastellm @Kourin1996

Reply:
cc @ferranbt, wdyt?

Reply:
> We shouldn't be using goto directives in Go in 2022, because we couldn't figure out how to write the program flow better.

I do not think we can generalize on the use of `goto`. Though I agree that It should not be used as a replacement for control loops, I think there are specific use cases where it is beneficial and we should not automatically discard it because of historical reasons (Note that many other projects from Hashicorp, Kubernetes, and Golang lang itself use `goto` statements as well).

I would prefer if your concern is about how `goto` is not being used correctly here and proposals to change the logic instead of an automatic discard.

Reply:
Anyway, if you feel strongly about this we can bring this topic on the code quality session and add it as part of the code guidelines.

Reply:
This should be solved case by case, it's still in use yes https://github.com/hashicorp/raft/search?q=goto but agree should be used with care, I expressed the same concern a while ago.